File: Src/Workspaces/VisualBasic/Formatting/Rules/AlignTokensFormattingRule.vb
' Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

Imports Microsoft.CodeAnalysis.Formatting.Rules
Imports Microsoft.CodeAnalysis.Options
Imports Microsoft.CodeAnalysis.VisualBasic.Syntax

Namespace Microsoft.CodeAnalysis.VisualBasic.Formatting
#If MEF Then
    <ExportFormattingRule(AlignTokensFormattingRule.Name, LanguageNames.VisualBasic)>
    <ExtensionOrder(After:=AdjustSpaceFormattingRule.Name)>
    Friend Class AlignTokensFormattingRule
#Else
    Friend Class AlignTokensFormattingRule
#End If
        Inherits BaseFormattingRule
        Friend Const Name As String = "VisualBasic Align Tokens Formatting Rule"

        Public Sub New()
        End Sub

        Public Overrides Sub AddAlignTokensOperations(operations As List(Of AlignTokensOperation), node As SyntaxNode, optionSet As OptionSet, nextOperation As NextAction(Of AlignTokensOperation))
            nextOperation.Invoke(operations)

            Dim queryExpression = TryCast(node, QueryExpressionSyntax)
            If queryExpression IsNot Nothing Then
                Dim tokens = New List(Of SyntaxToken)()
                tokens.AddRange(queryExpression.Clauses.Select(Function(q) q.GetFirstToken(includeZeroWidth:=True)))

                If tokens.Count > 1 Then
                    AddAlignIndentationOfTokensToBaseTokenOperation(operations, queryExpression, tokens(0), tokens.Skip(1))
                End If
                Return
            End If

            Dim multiLineLambda = TryCast(node, MultiLineLambdaExpressionSyntax)
            If multiLineLambda IsNot Nothing Then
                Dim baseToken = multiLineLambda.Begin.GetFirstToken(includeZeroWidth:=True)
                Dim tokenToAlign = multiLineLambda.End.GetFirstToken(includeZeroWidth:=True)

                AddAlignIndentationOfTokensToBaseTokenOperation(operations, multiLineLambda, baseToken, SpecializedCollections.SingletonEnumerable(tokenToAlign))
                Return
            End If

            Dim labelStatement = TryCast(node, LabelStatementSyntax)
            If labelStatement IsNot Nothing Then
                Dim nextToken = labelStatement.ColonToken.GetNextToken(includeZeroWidth:=True)
                AddAlignPositionOfTokensToIndentationOperation(operations, labelStatement, SpecializedCollections.SingletonEnumerable(nextToken))
            End If
        End Sub
    End Class
End Namespace

File: Src/Workspaces/CSharp/Formatting/Rules/AlignTokensFormattingRule.cs
// Copyright (c) Microsoft Open Technologies, Inc.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.

using System.Collections.Generic;
using Microsoft.CodeAnalysis.CSharp.Syntax;
using Microsoft.CodeAnalysis.Formatting.Rules;
using Microsoft.CodeAnalysis.Options;
using Roslyn.Utilities;

namespace Microsoft.CodeAnalysis.CSharp.Formatting
{
#if MEF
    [ExportFormattingRule(Name, LanguageNames.CSharp)]
#endif
    internal class AlignTokensFormattingRule : BaseFormattingRule
    {
        internal const string Name = "CSharp Align Tokens Formatting Rule";

        public override void AddAlignTokensOperations(List<AlignTokensOperation> list, SyntaxNode node, OptionSet optionSet, NextAction<AlignTokensOperation> nextOperation)
        {
            nextOperation.Invoke(list);
            var syntaxNode = node;

            var bracePair = node.GetBracePair();

            if (!bracePair.IsValidBracePair())
            {
                return;
            }

            if (syntaxNode.IsLambdaBodyBlock() ||
                syntaxNode.IsAnonymousMethodBlock() ||
                node is InitializerExpressionSyntax)
            {
                AddAlignIndentationOfTokensToFirstTokenOfBaseTokenLineOperation(list, syntaxNode, bracePair.Item1, SpecializedCollections.SingletonEnumerable((SyntaxToken)bracePair.Item2));
            }
        }
    }
}
